### PR TITLE
add missing arm64 node selector during update kube-proxy

### PR DESCRIPTION
**Pull request by aclevername**
### Description
https://github.com/weaveworks/eksctl/issues/2763

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by aclevername** [comment]
tested it by removing the `arm64` value manually and running an update.
Before:
```
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/os
                operator: In
                values:
                - linux
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
              - key: eks.amazonaws.com/compute-type
                operator: NotIn
                values:
                - fargate
```

After:
```
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/os
                operator: In
                values:
                - linux
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
                - arm64
              - key: eks.amazonaws.com/compute-type
                operator: NotIn
                values:
                - fargate
```

File: pkg/addons/default/kube_proxy.go [hunk 1]
```diff
@@ -97,6 +110,38 @@ func UpdateKubeProxyImageTag(clientSet kubernetes.Interface, controlPlaneVersion
 	return false, nil
 }
 
+func isMissingArm64NodeSelector(daemonSet *v1.DaemonSet) bool {
+	if daemonSet.Spec.Template.Spec.Affinity != nil &&
+		daemonSet.Spec.Template.Spec.Affinity.NodeAffinity != nil &&
+		daemonSet.Spec.Template.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
```
Comment (aclevername) [inline pkg/addons/default/kube_proxy.go:116]: :vomiting_face: I think this is unavoidable AFAIK

File: pkg/addons/default/kube_proxy.go [hunk 2]
```diff
@@ -44,7 +46,7 @@ func IsKubeProxyUpToDate(clientSet kubernetes.Interface, controlPlaneVersion str
 }
 
 // UpdateKubeProxyImageTag updates image tag for kube-system:daemonset/kube-proxy based to match controlPlaneVersion
```
Comment (Callisto13) [inline pkg/addons/default/kube_proxy.go:48]:
```suggestion
// UpdateKubeProxy updates image tag for kube-system:daemonset/kube-proxy based to match controlPlaneVersion
```

Comment (Callisto13) [inline pkg/addons/default/kube_proxy.go:48]:
```suggestion
// UpdateKubeProxy updates image tag for kube-system:daemonset/kube-proxy based to match controlPlaneVersion
```

File: pkg/addons/default/kube_proxy.go [hunk 3]
```diff
@@ -55,6 +57,12 @@ func UpdateKubeProxyImageTag(clientSet kubernetes.Interface, controlPlaneVersion
 		}
 		return false, errors.Wrapf(err, "getting %q", KubeProxy)
 	}
+
+	hasArm64NodeSelector := hasArm64NodeSelector(d)
+	if hasArm64NodeSelector {
+		logger.Info("missing arm64 nodeSelector value")
+	}
```
Comment (Callisto13) [inline pkg/addons/default/kube_proxy.go:64]: Either this is confusing or it's wrong. Should it be `if !hasArm64NodeSelector` then log "missing"?

File: pkg/addons/default/kube_proxy.go (hunk 1 above)
Comment (Callisto13) [reply pkg/addons/default/kube_proxy.go:116]: 😭

File: pkg/addons/default/kube_proxy.go (hunk 2 above)
Comment (Callisto13) [reply pkg/addons/default/kube_proxy.go:48]: how did i end up with 2 of these?

File: pkg/addons/default/kube_proxy.go [hunk 4]
```diff
@@ -55,6 +57,12 @@ func UpdateKubeProxyImageTag(clientSet kubernetes.Interface, controlPlaneVersion
 		}
 		return false, errors.Wrapf(err, "getting %q", KubeProxy)
 	}
+
+	hasArm64NodeSelector := daemeonSetHasArm64NodeSelector(d)
+	if !hasArm64NodeSelector {
+		logger.Info("missing arm64 nodeSelector value")
```
Comment (Callisto13) [inline pkg/addons/default/kube_proxy.go:63]: Does it make sense to log this down on 102 where we do the adding?

**Review by Callisto13 — APPROVED** [review]

File: pkg/addons/default/kube_proxy.go (hunk 4 above)
Comment (aclevername) [reply pkg/addons/default/kube_proxy.go:63]: we do a log first as we might exit earlier if `plan` is `false`

Comment (Callisto13) [reply pkg/addons/default/kube_proxy.go:63]: ah i see

**Pull request merged**